### ISSUE TITLE
Leak when finishing integration test

### ISSUE THREAD

**Issue by motxx** [hyperledger/iroha#782]
I tried `valgrind --leak-check=full ./build/test_bin/tx_pipeline_integration_test` at current develop and https://github.com/hyperledger/iroha/commit/792828fa2a6a69296960457fe89f85397361cd11 (before my rework integration test)
In both, I found some leak (in yac and rxcpp?):

```
[       OK ] TxPipelineIntegrationTest.TxPipelineTest (20233 ms)
[----------] 1 test from TxPipelineIntegrationTest (20243 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (20289 ms total)
[  PASSED  ] 1 test.
==9828==
==9828== HEAP SUMMARY:
==9828==     in use at exit: 93,940 bytes in 332 blocks
==9828==   total heap usage: 20,476 allocs, 20,144 frees, 10,115,776 bytes allocated
==9828==
==9828== 1,704 (128 direct, 1,576 indirect) bytes in 1 blocks are definitely lost in loss record 74 of 77
==9828==    at 0x4C2E216: operator new(unsigned long) (vg_replace_malloc.c:334)
==9828==    by 0x47DB23: __gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> >::allocate(unsigned long, void const*) (new_allocator.h:104)
==9828==    by 0x47AF4C: std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> > >::allocate(std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> >&, unsigned long) (alloc_traits.h:491)
==9828==    by 0x47690B: std::__allocated_ptr<std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> > > std::__allocate_guarded<std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> > >(std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> >&) (allocated_ptr.h:102)
==9828==    by 0x46F8E0: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::_Sp_make_shared_tag, rxcpp::detail::composite_subscription_inner::composite_subscription_state*, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr_base.h:615)
==9828==    by 0x46A127: std::__shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::_Sp_make_shared_tag, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr_base.h:1097)
==9828==    by 0x462505: std::shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state>::shared_ptr<std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::_Sp_make_shared_tag, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr.h:319)
==9828==    by 0x457875: std::shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state> std::allocate_shared<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr.h:620)
==9828==    by 0x44F9E6: std::shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state> std::make_shared<rxcpp::detail::composite_subscription_inner::composite_subscription_state>() (shared_ptr.h:636)
==9828==    by 0x445A76: rxcpp::detail::composite_subscription_inner::composite_subscription_inner() (rx-subscription.hpp:296)
==9828==    by 0x445DDB: rxcpp::composite_subscription::composite_subscription() (rx-subscription.hpp:382)
==9828==    by 0x509417: rxcpp::subjects::subject<iroha::consensus::yac::CommitMessage>::subject() (rx-subject.hpp:245)
==9828==
==9828== LEAK SUMMARY:
==9828==    definitely lost: 128 bytes in 1 blocks
==9828==    indirectly lost: 1,576 bytes in 20 blocks
==9828==      possibly lost: 0 bytes in 0 blocks
==9828==    still reachable: 92,236 bytes in 311 blocks
==9828==         suppressed: 0 bytes in 0 blocks
==9828== Reachable blocks (those to which a pointer was found) are not shown.
==9828== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==9828==
==9828== For counts of detected and suppressed errors, rerun with: -v
==9828== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

All logs at: https://gist.github.com/motxx/3bb288501eb0b22d5001afdec073fd75

**Comment by l4l** [comment]
I have even worse results:
```==17594== 
==17594== HEAP SUMMARY:
==17594==     in use at exit: 25,149 bytes in 328 blocks
==17594==   total heap usage: 20,216 allocs, 19,888 frees, 10,177,115 bytes allocated
==17594== 
==17594== Thread 1:
==17594== 1,704 (128 direct, 1,576 indirect) bytes in 1 blocks are definitely lost in loss record 153 of 154
==17594==    at 0x4C2D52F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17594==    by 0x1982D9: __gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> >::allocate(unsigned long, void const*) (new_allocator.h:111)
==17594==    by 0x1952A4: std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> > >::allocate(std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> >&, unsigned long) (alloc_traits.h:436)
==17594==    by 0x1904D9: std::__allocated_ptr<std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> > > std::__allocate_guarded<std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> > >(std::allocator<std::_Sp_counted_ptr_inplace<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>, (__gnu_cxx::_Lock_policy)2> >&) (allocated_ptr.h:104)
==17594==    by 0x188F22: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::_Sp_make_shared_tag, rxcpp::detail::composite_subscription_inner::composite_subscription_state*, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr_base.h:635)
==17594==    by 0x18367F: std::__shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::_Sp_make_shared_tag, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr_base.h:1295)
==17594==    by 0x17BAD7: std::shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state>::shared_ptr<std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::_Sp_make_shared_tag, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr.h:344)
==17594==    by 0x1713EF: std::shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state> std::allocate_shared<rxcpp::detail::composite_subscription_inner::composite_subscription_state, std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state>>(std::allocator<rxcpp::detail::composite_subscription_inner::composite_subscription_state> const&) (shared_ptr.h:691)
==17594==    by 0x169255: std::shared_ptr<rxcpp::detail::composite_subscription_inner::composite_subscription_state> std::make_shared<rxcpp::detail::composite_subscription_inner::composite_subscription_state>() (shared_ptr.h:707)
==17594==    by 0x15FA3C: rxcpp::detail::composite_subscription_inner::composite_subscription_inner() (rx-subscription.hpp:296)
==17594==    by 0x15FDA3: rxcpp::composite_subscription::composite_subscription() (rx-subscription.hpp:382)
==17594==    by 0x22043F: rxcpp::subjects::subject<iroha::consensus::yac::CommitMessage>::subject() (rx-subject.hpp:245)
==17594== 
==17594== 4,096 bytes in 1 blocks are definitely lost in loss record 154 of 154
==17594==    at 0x4C2CE5F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==17594==    by 0xAD99E18: ???
==17594==    by 0xADB53C1: ???
==17594==    by 0xADABA41: ???
==17594==    by 0xADAC1B1: ???
==17594==    by 0x6F06C6C: gethostbyname2_r@@GLIBC_2.2.5 (in /usr/lib/libc-2.26.so)
==17594==    by 0x6EDE935: gaih_inet.constprop.7 (in /usr/lib/libc-2.26.so)
==17594==    by 0x6EDF573: getaddrinfo (in /usr/lib/libc-2.26.so)
==17594==    by 0x4DDECE: tacopie::tcp_socket::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int) (tcp_socket.cpp:135)
==17594==    by 0x4D88A6: tacopie::tcp_client::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int) (tcp_client.cpp:79)
==17594==    by 0x4D050F: connect (tcp_client.hpp:44)
==17594==    by 0x4D050F: cpp_redis::network::redis_connection::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, std::function<void (cpp_redis::network::redis_connection&)> const&, std::function<void (cpp_redis::network::redis_connection&, cpp_redis::reply&)> const&) (redis_connection.cpp:61)
==17594==    by 0x4BA7F3: cpp_redis::redis_client::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, std::function<void (cpp_redis::redis_client&)> const&) (redis_client.cpp:51)
==17594== 
==17594== LEAK SUMMARY:
==17594==    definitely lost: 4,224 bytes in 2 blocks
==17594==    indirectly lost: 1,576 bytes in 20 blocks
==17594==      possibly lost: 0 bytes in 0 blocks
==17594==    still reachable: 19,349 bytes in 306 blocks
==17594==         suppressed: 0 bytes in 0 blocks
```

Ping @grimadas

**Comment by neewy** [comment]
Why Bulat? Can you explain the output of valgrind? Where the problem is localized?

**Comment by l4l** [comment]
Should be fixed in #883, I have only false-positives right now
